text/0000-array-private-classes.md

```diff
@@ -0,0 +1,366 @@
+found some tangible benefit in doing so), then it would not be a breaking change
+at all, because the new proposed type would be a subtype of the existing return
+types.
+
+It'll always be easier to get more specific with these return types than more
+general. So we should default to offering the most general return type that is
+still useful to users, and ratchet down to more specific subtypes as needed (as
+tangible use cases for more specific return types are discovered). Ratcheting down
+will always be possible without a breaking change, but ratcheting up will be a
+breaking change each time. It's best to make such "ratchet up" breaking changes
+like this RFC before we reach Pony 1.0.0 as part of a general effort to make Pony
+and its standard library stable.
+
+# Detailed design
+
+Iterating functions in collections `keys`, `values` and `pairs` are changed to
+return `Iterator` and the classes that implement these iterators are made private.
+Here are the full implementation of these functions for the `Array` class (changes
+in other collection classes are identical).
+
+As the function `values` of class `Array` uses an iterator with a `rewind` function
+that is not part of the `Iterator` interface, a new interface `CollectionIterator`
+is added to enable creation of rewindable iterators.
+
+```pony
+  fun keys(): CollectionIterator[USize]^ =>
+    """
+    Return an iterator over the indices in the array.
+    """
+    _ArrayKeys[A, this->Array[A]](this)
+
+  fun values(): CollectionIterator[this->A]^ =>
+    """
+    Return an iterator over the values in the array.
+    """
+    _ArrayValues[A, this->Array[A]](this)
+
+  fun pairs(): CollectionIterator[(USize, this->A)]^ =>
+    """
+    Return an iterator over the (index, value) pairs in the array.
+    """
+    _ArrayPairs[A, this->Array[A]](this)
+```
+
+Note: To remain consistent with `Array` behaviour, functions `keys` and `pairs`
+will return a `CollectionIterator`.
+
+```pony
+interface CollectionIterator[A] is Iterator[A]
+  """
+  A `CollectionIterator` is an iterator that can be rewinded, that is start
+  again from first item. The data structure being iterated on can't change the
+  order it return iterated items.
+  """
+  fun has_next(): Bool
+    """
+    Return `true` when function `next` can be called to get next iteration item.
+    """
+
+  fun ref next(): A ?
+    """
+    Return the next item of the iteration or an error in case there are no other
+    items. A previous call to `has_next` check if we can continue iteration.
+    """
+
+  fun ref rewind(): Iterator[A]^
+    """
+    Start the iterator over again from the beginning.
+    """
+```
+
+The code of the standard library is adapted to remove use of these now private
+classes, mainly in tests. Here are the files that must be changed:
+
+* `packages/builtin/array.pony` as shown above
+* `packages/itertools/iter.pony` in function `cycle`
```

Comment:
I see that in source `cycle` has a type annotation to `ArrayValues[A!, Array[A!]]`. This would indeed need to change, but does clarify by previous question/concern of where `itertools` needs to change.